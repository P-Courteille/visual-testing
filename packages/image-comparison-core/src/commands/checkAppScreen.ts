import { screenMethodCompareOptions } from '../helpers/options.js'
import type {  ImageCompareOptions, ImageCompareResult } from '../methods/images.interfaces.js'
import { executeImageCompare } from '../methods/images.js'
import { determineDeviceBlockOuts, determineIgnoreRegions } from '../methods/rectangles.js'
import type { InternalCheckScreenMethodOptions } from './check.interfaces.js'
import saveAppScreen from './saveAppScreen.js'
import type { ElementIgnore } from './element.interfaces.js'

/**
 * Compare an image of the viewport of the screen
 */
export default async function checkAppScreen(
    {
        browserInstance,
        checkScreenOptions,
        folders,
        instanceData,
        isNativeContext = true,
        tag,
        testContext,
    }: InternalCheckScreenMethodOptions
): Promise<ImageCompareResult | number> {
    // 1. Set some vars
    const saveAppScreenOptions = {
        wic: checkScreenOptions.wic,
        method:{
            ...{ hideElements: [] },
            ...{ removeElements:  [] },
        }
    }
    const screenCompareOptions = {
        ...checkScreenOptions.wic.compareOptions,
        ...checkScreenOptions.method,
        // Use the hide and remove elements from the checkScreenOptions and add them to the ignore array
        ignore: [
            ...checkScreenOptions.method.ignore || [],
            ...checkScreenOptions.method.hideElements as unknown as ElementIgnore[] || [],
            ...checkScreenOptions.method.removeElements as unknown as ElementIgnore[] || [],
        ]

    }
    const { isAndroid, isMobile, deviceRectangles, browserName, deviceName, nativeWebScreenshot: isAndroidNativeWebScreenshot } = instanceData
    const { actualFolder, baselineFolder, diffFolder } = folders
    const { autoSaveBaseline, savePerInstance, compareOptions } = checkScreenOptions.wic

    // 2. Take the actual screenshot and retrieve the needed data
    const { devicePixelRatio, fileName } = await saveAppScreen({
        browserInstance,
        folders,
        instanceData,
        isNativeContext,
        saveScreenOptions: saveAppScreenOptions,
        tag,
    })

    // 3. Determine the ignore regions
    const ignoreRegions = await determineIgnoreRegions(browserInstance, screenCompareOptions.ignore || [])
    const deviceIgnoreRegions = await determineDeviceBlockOuts({
        isAndroid,
        screenCompareOptions,
        instanceData,
    })

    // 4a. Determine the compare options
    const methodCompareOptions = screenMethodCompareOptions(checkScreenOptions.method)

    const executeCompareOptions: ImageCompareOptions = {
        compareOptions: {
            wic: compareOptions,
            method: methodCompareOptions,
        },
        devicePixelRatio,
        deviceRectangles,
        fileName,
        folderOptions: {
            autoSaveBaseline,
            actualFolder,
            baselineFolder,
            diffFolder,
            browserName,
            deviceName,
            isMobile,
            savePerInstance,
        },
        ignoreRegions: [...ignoreRegions, ...deviceIgnoreRegions],
        isAndroid,
        isAndroidNativeWebScreenshot,
    }

    // 4b Now execute the compare and return the data
    return executeImageCompare({
        isViewPortScreenshot: true,
        isNativeContext,
        options: executeCompareOptions,
        testContext,
    })
}
