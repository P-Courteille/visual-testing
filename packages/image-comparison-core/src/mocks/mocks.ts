import type { BeforeScreenshotOptions } from '../helpers/beforeScreenshot.interfaces.js'

export const BEFORE_SCREENSHOT_OPTIONS: BeforeScreenshotOptions = {
    instanceData: {
        appName: 'chrome-app',
        browserName: 'chrome',
        browserVersion: '75.0.1',
        deviceName: '',
        devicePixelRatio: 1,
        initialDevicePixelRatio: 1,
        deviceRectangles: {
            statusBar: { y: 0, x: 0, width: 0, height: 0 },
            homeBar: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 0, width: 0 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 0, height: 0 },
            viewport: { y: 0, x: 0, width: 0, height: 0 },
            bottomBar: { y: 0, x: 0, width: 0, height: 0 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
        },
        isAndroid: false,
        isIOS: false,
        isMobile: false,
        logName: 'chrome-latest',
        name: 'chrome-name',
        nativeWebScreenshot: true,
        platformName: 'Windows 10',
        platformVersion: '1234',
    },
    addressBarShadowPadding: 6,
    disableBlinkingCursor: true,
    disableCSSAnimation: true,
    enableLayoutTesting: false,
    noScrollBars: true,
    toolBarShadowPadding: 6,
    hideElements: [<HTMLElement>(<any>'<div></div>')],
    removeElements: [<HTMLElement>(<any>'<div></div>')],
    waitForFontsLoaded: true,
}
export const CONFIGURABLE = {
    writable: true,
    configurable: true,
}
export const NAVIGATOR_APP_VERSIONS = {
    ANDROID: {
        7: '5.0 (Linux; Android 7.1.1; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        8: '5.0 (Linux; Android 8.1; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        9: '5.0 (Linux; Android 9; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        10: '5.0 (Linux; Android 10; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
        11: '5.0 (Linux; Android 11; Android SDK built for x86 Build/NYC) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/55.0.2883.91 Mobile Safari/537.3',
    },
    IOS: {
        10: '5.0 (iPhone; CPU iPhone OS 10_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/10.0 Mobile/15E148 Safari/604.1',
        11: '5.0 (iPhone; CPU iPhone OS 11_4 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/11.0 Mobile/15E148 Safari/604.1',
        12: '5.0 (iPhone; CPU iPhone OS 12_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/12.0 Mobile/15E148 Safari/604.1',
        13: '5.0 (iPhone; CPU iPhone OS 13_3 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0 Mobile/15E148 Safari/604.1',
        14: '5.0 (iPhone; CPU iPhone OS 14_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Mobile/15E148 Safari/604.1',
        15: '5.0 (iPhone; CPU iPhone OS 15_0 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Mobile/15E148 Safari/604.1',
    },
    IPADOS: {
        13: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/13.0.1 Safari/605.1.15',
        14: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0 Safari/605.1.15',
        15: 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.0 Safari/605.1.15',
    },
}
export const ANDROID_DEVICES = {
    NEXUS_5X: {
        height: 732,
        width: 412,
        innerHeight: 604,
        innerWidth: 412,
    },
    NEXUS_5X_INNER_HEIGHT: {
        height: 732,
        width: 412,
        innerHeight: 800,
        innerWidth: 412,
    },
    TABLET_WIDTH: {
        height: 768,
        width: 1024,
        innerHeight: 604,
        innerWidth: 412,
    },
}
export const IOS_DEVICES = {
    IPHONE: {
        height: 667,
        width: 375,
        innerHeight: 553,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPHONE_X: {
        height: 812,
        width: 375,
        innerHeight: 635,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPHONE_HEIGHT: {
        height: 896,
        width: 1024,
        innerHeight: 719,
        innerWidth: 414,
        scrollWidth: 414,
        sideBar: 0,
    },
    IPHONE_11: {
        height: 896,
        width: 375,
        innerHeight: 635,
        innerWidth: 375,
        scrollWidth: 375,
        sideBar: 0,
    },
    IPAD: {
        height: 1366,
        width: 1024,
        innerHeight: 1292,
        innerWidth: 1024,
        scrollWidth: 1024,
        sideBar: 0,
    },
    IPAD_LANDSCAPE: {
        height: 1366,
        width: 1024,
        innerHeight: 746,
        innerWidth: 1046,
        scrollWidth: 1046,
        sideBar: 320,
    },
    IPAD_BIG_SIZE: {
        height: 5432,
        width: 9876,
        innerHeight: 5324,
        innerWidth: 9768,
        scrollWidth: 9768,
        sideBar: 108,
    },
    IPAD_PRO_LANDSCAPE: {
        height: 1366,
        width: 1024,
        innerHeight: 954,
        innerWidth: 1046,
        scrollWidth: 1046,
        sideBar: 320,
    },
}

export const BASE_CHECK_OPTIONS = {
    wic: {
        addressBarShadowPadding: 6,
        autoElementScroll: true,
        addIOSBezelCorners: false,
        autoSaveBaseline: false,
        clearFolder: false,
        userBasedFullPageScreenshot: false,
        enableLegacyScreenshotMethod: false,
        formatImageName: '{tag}-{logName}-{width}x{height}-dpr-{dpr}',
        isHybridApp: false,
        savePerInstance: true,
        toolBarShadowPadding: 6,
        disableBlinkingCursor: false,
        disableCSSAnimation: false,
        enableLayoutTesting: false,
        fullPageScrollTimeout: 1500,
        hideScrollBars: true,
        waitForFontsLoaded: true,
        compareOptions: {
            ignoreAlpha: false,
            ignoreAntialiasing: false,
            ignoreColors: false,
            ignoreLess: false,
            ignoreNothing: false,
            rawMisMatchPercentage: false,
            returnAllCompareData: false,
            saveAboveTolerance: 0,
            scaleImagesToSameSize: false,
            blockOutSideBar: false,
            blockOutStatusBar: false,
            blockOutToolBar: false,
            createJsonReportFiles: false,
            diffPixelBoundingBoxProximity: 5,
        },
        tabbableOptions: {
            circle: {
                backgroundColor: 'rgba(255, 0, 0, 0.4)',
                borderColor: 'rgba(255, 0, 0, 1)',
                borderWidth: 1,
                fontColor: 'rgba(0, 0, 0, 1)',
                fontFamily: 'Arial',
                fontSize: 10,
                size: 10,
            },
            line: {
                color: 'rgba(255, 0, 0, 1)',
                width: 1,
            },
        }
    },
    instanceData: {
        appName: 'TestApp',
        browserName: 'Chrome',
        browserVersion: '118.0.0.0',
        deviceName: 'iPhone 14',
        devicePixelRatio: 2,
        deviceRectangles: {
            bottomBar: { y: 800, x: 0, width: 390, height: 0 },
            homeBar: { x: 0, y: 780, width: 390, height: 34 },
            leftSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            rightSidePadding: { y: 0, x: 0, width: 0, height: 0 },
            screenSize: { height: 844, width: 390 },
            statusBar: { x: 0, y: 0, width: 390, height: 47 },
            statusBarAndAddressBar: { y: 0, x: 0, width: 390, height: 47 },
            viewport: { y: 47, x: 0, width: 390, height: 733 }
        },
        initialDevicePixelRatio: 2,
        isAndroid: false,
        isIOS: true,
        isMobile: true,
        logName: 'test-log',
        name: 'test-device',
        nativeWebScreenshot: false,
        platformName: 'iOS',
        platformVersion: '17.0'
    },
    folders: {
        actualFolder: '/test/actual',
        baselineFolder: '/test/baseline',
        diffFolder: '/test/diff'
    },
    testContext: {
        commandName: 'checkScreen',
        framework: 'vitest',
        parent: 'test suite',
        title: 'test title',
        tag: 'test-tag',
        instanceData: {
            browser: { name: 'Chrome', version: '118.0.0.0' },
            deviceName: 'iPhone 14',
            platform: { name: 'iOS', version: '17.0' },
            app: 'TestApp',
            isMobile: true,
            isAndroid: false,
            isIOS: true
        }
    }
}
