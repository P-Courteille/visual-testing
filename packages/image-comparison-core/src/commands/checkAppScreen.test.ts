import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest'
import checkAppScreen from './checkAppScreen.js'
import type { InternalCheckScreenMethodOptions } from './check.interfaces.js'
import { BASE_CHECK_OPTIONS } from '../mocks/mocks.js'

// Mock the dependencies
vi.mock('../helpers/options.js', () => ({
    screenMethodCompareOptions: vi.fn().mockReturnValue({
        ignoreAlpha: false,
        ignoreAntialiasing: false,
        ignoreColors: false,
        ignoreLess: false,
        ignoreNothing: false,
        rawMisMatchPercentage: false,
        returnAllCompareData: false,
        saveAboveTolerance: 0,
        scaleImagesToSameSize: false,
    })
}))

vi.mock('../methods/images.js', () => ({
    executeImageCompare: vi.fn().mockResolvedValue({
        fileName: 'test-result.png',
        misMatchPercentage: 0,
        isExactSameImage: true,
        isNewBaseline: false,
        isAboveTolerance: false,
    })
}))

vi.mock('../methods/rectangles.js', () => ({
    determineIgnoreRegions: vi.fn().mockResolvedValue([
        { x: 0, y: 0, width: 100, height: 100 }
    ]),
    determineDeviceBlockOuts: vi.fn().mockResolvedValue([
        { x: 0, y: 0, width: 50, height: 50 }
    ])
}))

vi.mock('./saveAppScreen.js', () => ({
    default: vi.fn().mockResolvedValue({
        devicePixelRatio: 2,
        fileName: 'test-screen.png'
    })
}))

describe('checkAppScreen', () => {
    let executeImageCompareSpy: ReturnType<typeof vi.fn>
    let saveAppScreenSpy: ReturnType<typeof vi.fn>
    let determineIgnoreRegionsSpy: ReturnType<typeof vi.fn>
    let determineDeviceBlockOutsSpy: ReturnType<typeof vi.fn>

    const baseOptions: InternalCheckScreenMethodOptions = {
        checkScreenOptions: {
            wic: BASE_CHECK_OPTIONS.wic,
            method: {
                hideElements: [],
                removeElements: [],
                ignore: [],
                blockOut: [],
                blockOutSideBar: false,
                blockOutStatusBar: false,
                blockOutToolBar: false,
                ignoreAlpha: false,
                ignoreAntialiasing: false,
                ignoreColors: false,
                ignoreLess: false,
                ignoreNothing: false,
                rawMisMatchPercentage: false,
                returnAllCompareData: false,
                saveAboveTolerance: 0,
                scaleImagesToSameSize: false,
            }
        },
        browserInstance: { isAndroid: false } as any,
        folders: BASE_CHECK_OPTIONS.folders,
        instanceData: BASE_CHECK_OPTIONS.instanceData,
        isNativeContext: true,
        tag: 'test-screen',
        testContext: BASE_CHECK_OPTIONS.testContext
    }

    beforeEach(async () => {
        const { executeImageCompare } = await import('../methods/images.js')
        const { determineIgnoreRegions, determineDeviceBlockOuts } = await import('../methods/rectangles.js')
        const saveAppScreen = (await import('./saveAppScreen.js')).default

        executeImageCompareSpy = vi.mocked(executeImageCompare)
        determineIgnoreRegionsSpy = vi.mocked(determineIgnoreRegions)
        determineDeviceBlockOutsSpy = vi.mocked(determineDeviceBlockOuts)
        saveAppScreenSpy = vi.mocked(saveAppScreen)
    })

    afterEach(() => {
        vi.clearAllMocks()
    })

    it('should execute checkAppScreen with basic options', async () => {
        const result = await checkAppScreen(baseOptions)

        expect(result).toMatchSnapshot()
        expect(saveAppScreenSpy).toHaveBeenCalledWith({
            browserInstance: baseOptions.browserInstance,
            folders: baseOptions.folders,
            instanceData: baseOptions.instanceData,
            isNativeContext: true,
            saveScreenOptions: {
                wic: baseOptions.checkScreenOptions.wic,
                method: {
                    hideElements: [],
                    removeElements: []
                }
            },
            tag: 'test-screen',
        })
    })

    it('should handle ignore regions and device blockouts', async () => {
        const mockElement = { elementId: 'test-element', selector: '#test' } as any
        const options = {
            ...baseOptions,
            checkScreenOptions: {
                ...baseOptions.checkScreenOptions,
                method: {
                    ...baseOptions.checkScreenOptions.method,
                    hideElements: [mockElement],
                    removeElements: [mockElement],
                    ignore: [mockElement]
                }
            }
        }

        await checkAppScreen(options)

        expect(determineIgnoreRegionsSpy).toHaveBeenCalledWith(
            options.browserInstance,
            [mockElement, mockElement, mockElement]
        )

        expect(determineDeviceBlockOutsSpy).toHaveBeenCalledWith({
            isAndroid: false,
            screenCompareOptions: expect.objectContaining({
                ignore: [mockElement, mockElement, mockElement]
            }),
            instanceData: options.instanceData
        })

        expect(executeImageCompareSpy).toHaveBeenCalledWith({
            options: expect.objectContaining({
                ignoreRegions: [
                    { x: 0, y: 0, width: 100, height: 100 },
                    { x: 0, y: 0, width: 50, height: 50 }
                ]
            }),
            testContext: expect.any(Object),
            isViewPortScreenshot: true,
            isNativeContext: true,
        })
    })

    it('should handle Android device correctly', async () => {
        const options = {
            ...baseOptions,
            browserInstance: { isAndroid: true } as any,
            instanceData: {
                ...baseOptions.instanceData,
                deviceName: 'Pixel 4',
                isAndroid: true,
                isIOS: false,
                platformName: 'Android',
                platformVersion: '11.0'
            },
            testContext: {
                ...baseOptions.testContext,
                instanceData: {
                    ...baseOptions.testContext.instanceData,
                    deviceName: 'Pixel 4',
                    platform: { name: 'Android', version: '11.0' },
                    isAndroid: true,
                    isIOS: false
                }
            }
        }

        await checkAppScreen(options)

        expect(determineDeviceBlockOutsSpy).toHaveBeenCalledWith({
            isAndroid: true,
            screenCompareOptions: expect.any(Object),
            instanceData: expect.objectContaining({
                isAndroid: true,
                platformName: 'Android'
            })
        })

        expect(executeImageCompareSpy).toHaveBeenCalledWith({
            options: expect.objectContaining({
                isAndroid: true,
                isAndroidNativeWebScreenshot: false
            }),
            testContext: expect.any(Object),
            isViewPortScreenshot: true,
            isNativeContext: true,
        })
    })

    it('should merge compare options correctly', async () => {
        const options = {
            ...baseOptions,
            checkScreenOptions: {
                ...baseOptions.checkScreenOptions,
                wic: {
                    ...baseOptions.checkScreenOptions.wic,
                    compareOptions: {
                        ...baseOptions.checkScreenOptions.wic.compareOptions,
                        ignoreAlpha: true,
                        ignoreAntialiasing: true,
                        ignoreColors: true,
                    }
                },
                method: {
                    ...baseOptions.checkScreenOptions.method,
                    ignoreAlpha: false,
                    ignoreAntialiasing: false,
                    ignoreColors: false,
                }
            }
        }

        await checkAppScreen(options)

        expect(executeImageCompareSpy).toHaveBeenCalledWith({
            options: expect.objectContaining({
                compareOptions: {
                    wic: expect.objectContaining({
                        ignoreAlpha: true,
                        ignoreAntialiasing: true,
                        ignoreColors: true,
                    }),
                    method: expect.objectContaining({
                        ignoreAlpha: false,
                        ignoreAntialiasing: false,
                        ignoreColors: false,
                    })
                }
            }),
            testContext: expect.any(Object),
            isViewPortScreenshot: true,
            isNativeContext: true,
        })
    })

    it('should spread hideElements and removeElements into ignore array', async () => {
        const mockElement1 = { elementId: 'hide-element', selector: '#hide' } as any
        const mockElement2 = { elementId: 'remove-element', selector: '#remove' } as any
        const mockElement3 = { elementId: 'ignore-element', selector: '#ignore' } as any

        const options = {
            ...baseOptions,
            checkScreenOptions: {
                ...baseOptions.checkScreenOptions,
                method: {
                    ...baseOptions.checkScreenOptions.method,
                    hideElements: [mockElement1],
                    removeElements: [mockElement2],
                    ignore: [mockElement3]
                }
            }
        }

        await checkAppScreen(options)

        expect(determineIgnoreRegionsSpy).toHaveBeenCalledWith(
            options.browserInstance,
            [mockElement3, mockElement1, mockElement2]
        )

        expect(executeImageCompareSpy).toHaveBeenCalledWith({
            options: expect.objectContaining({
                ignoreRegions: [
                    { x: 0, y: 0, width: 100, height: 100 },
                    { x: 0, y: 0, width: 50, height: 50 }
                ]
            }),
            testContext: expect.any(Object),
            isViewPortScreenshot: true,
            isNativeContext: true,
        })
    })

    it('should create screenCompareOptions with correct structure', async () => {
        const mockElement1 = { elementId: 'hide-element', selector: '#hide' } as any
        const mockElement2 = { elementId: 'remove-element', selector: '#remove' } as any
        const mockElement3 = { elementId: 'ignore-element', selector: '#ignore' } as any

        const options = {
            ...baseOptions,
            checkScreenOptions: {
                ...baseOptions.checkScreenOptions,
                wic: {
                    ...baseOptions.checkScreenOptions.wic,
                    compareOptions: {
                        ...baseOptions.checkScreenOptions.wic.compareOptions,
                        ignoreAlpha: true,
                        ignoreAntialiasing: true,
                        ignoreColors: true,
                    }
                },
                method: {
                    hideElements: [mockElement1],
                    removeElements: [mockElement2],
                    ignore: [mockElement3],
                    ignoreAlpha: false,
                    ignoreAntialiasing: false,
                    ignoreColors: false,
                }
            }
        }

        await checkAppScreen(options)

        expect(determineIgnoreRegionsSpy).toHaveBeenCalledWith(
            options.browserInstance,
            [mockElement3, mockElement1, mockElement2]
        )

        expect(determineDeviceBlockOutsSpy).toHaveBeenCalledWith({
            isAndroid: false,
            screenCompareOptions: expect.objectContaining({
                hideElements: [mockElement1],
                removeElements: [mockElement2],
                ignore: [mockElement3, mockElement1, mockElement2],
                ignoreAlpha: false,
                ignoreAntialiasing: false,
                ignoreColors: false,
            }),
            instanceData: options.instanceData
        })
    })

    it('should spread wic.compareOptions and method options into screenCompareOptions', async () => {
        const options = {
            ...baseOptions,
            checkScreenOptions: {
                ...baseOptions.checkScreenOptions,
                wic: {
                    ...baseOptions.checkScreenOptions.wic,
                    compareOptions: {
                        ignoreAlpha: true,
                        ignoreAntialiasing: true,
                        ignoreColors: true,
                        blockOutSideBar: true,
                        blockOutStatusBar: true,
                        blockOutToolBar: true,
                        createJsonReportFiles: true,
                        diffPixelBoundingBoxProximity: 10,
                        ignoreLess: true,
                        ignoreNothing: true,
                        rawMisMatchPercentage: true,
                        returnAllCompareData: true,
                        saveAboveTolerance: 1,
                        scaleImagesToSameSize: true,
                    }
                },
                method: {
                    ignoreAlpha: false,
                    ignoreAntialiasing: false,
                    ignoreColors: false,
                    blockOutSideBar: false,
                    blockOutStatusBar: false,
                    blockOutToolBar: false,
                    createJsonReportFiles: false,
                    diffPixelBoundingBoxProximity: 5,
                    ignoreLess: false,
                    ignoreNothing: false,
                    rawMisMatchPercentage: false,
                    returnAllCompareData: false,
                    saveAboveTolerance: 0,
                    scaleImagesToSameSize: false,
                }
            }
        }

        await checkAppScreen(options)

        expect(determineDeviceBlockOutsSpy).toHaveBeenCalledWith({
            isAndroid: false,
            screenCompareOptions: expect.objectContaining({
                // Method options should override wic options
                ignoreAlpha: false,
                ignoreAntialiasing: false,
                ignoreColors: false,
                blockOutSideBar: false,
                blockOutStatusBar: false,
                blockOutToolBar: false,
                createJsonReportFiles: false,
                diffPixelBoundingBoxProximity: 5,
                ignoreLess: false,
                ignoreNothing: false,
                rawMisMatchPercentage: false,
                returnAllCompareData: false,
                saveAboveTolerance: 0,
                scaleImagesToSameSize: false,
            }),
            instanceData: options.instanceData
        })
    })
})
