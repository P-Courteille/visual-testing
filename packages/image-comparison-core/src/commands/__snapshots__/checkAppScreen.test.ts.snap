// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`checkAppScreen > should execute checkAppScreen with basic options 1`] = `
{
  "fileName": "test-result.png",
  "isAboveTolerance": false,
  "isExactSameImage": true,
  "isNewBaseline": false,
  "misMatchPercentage": 0,
}
`;
